### ISSUE TITLE
[Bounty DE-SCOPED] ttnn.exp2: fp32 path returns NaN / finite garbage instead of +inf for x >= 128 (missing overflow guard)

### ISSUE THREAD

**Issue by 5awtooth** [tenstorrent/tt-metal#54054]
## Summary

`ttnn.exp2`'s fp32-accurate kernel (`_sfpu_exp2_fp32_accurate_`, selected when
`is_fp32_dest_acc_en == true`) reconstructs `2^x = 2^round(x) * 2^f` by evaluating a minimax
polynomial `r = 2^f` (`f = x - round(x)`) and then rebuilding the exponent with
`setexp(r, e)`, `e = biased_exp(r) + round(x)`.

The exponent path has an **underflow** guard (`e < 1 -> 0`) but **no overflow guard**. For
`x >= 128` the true result overflows fp32 and should be `+inf`. Instead `e` reaches `>= 255`
and the *unconditional* `setexp(r, e)` writes an out-of-range exponent field:

* `x` in `[128, 128.5)` -> `e == 255` with a nonzero mantissa -> **NaN**
* `x >= 128.5` -> `e >= 256`, `setexp` keeps only the low 8 bits -> **wrapped finite garbage**

The kernel even **precomputes the correct overflow result** one line earlier
(`y *= std::numeric_limits<float>::infinity();`, i.e. `y = r * inf = +inf`) and then throws it
away by overwriting `y` with the `setexp` result. The overflow half of the reconstruction is
simply missing.

`exp2(x)` for `x >= 128` is unambiguously `+inf` (2^128 > FLT_MAX), so this is never a
representable-range question — the correct answer exists (`+inf`), is cheap, and is already
computed. Worst of all, the `x >= 128.5` cases return an *ordinary-looking finite number* that
propagates silently: `exp2(256)` returns `1.0`, `exp2(255)` returns `0.5`, `exp2(160)` returns
`1.26e-29`.

## Measured (source-faithful fp32 model of the kernel vs `numpy.exp2` fp32 golden)

The table below comes from a bit-exact fp32 reimplementation of the kernel arithmetic
(sign-magnitude rounding of `round(x)`, the exact minimax coefficients, `exexp`, `e = exexp+round(x)`,
`setexp` = low-8-bits-of-`e` into the exponent field). Reproduction script attached
(`verify_exp2.py`, `proof.txt`); values should also reproduce on ttsim / silicon.

| x        | kernel (fp32 path) | true `exp2(x)` fp32 |     |
|----------|--------------------|---------------------|-----|
| 127.0    | 1.70141e+38        | 1.70141e+38         | ok  |
| 127.5    | 2.40616e+38        | 2.40616e+38         | ok  |
| 128.0    | `inf`              | `inf`               | ok (accidental: f = 0) |
| 128.01   | `NaN`              | `inf`               | ✗   |
| 128.25   | `NaN`              | `inf`               | ✗   |
| 129.0    | 0.0                | `inf`               | ✗   |
| 160.0    | 1.26218e-29        | `inf`               | ✗   |
| 200.0    | 1.38778e-17        | `inf`               | ✗   |
| 255.0    | 0.5                | `inf`               | ✗   |
| 256.0    | 1.0                | `inf`               | ✗   |
| 300.0    | 1.75922e+13        | `inf`               | ✗   |

The threshold is sharp and it is exactly the fp32 overflow point of `exp2`:

* every `x < 128` is correct (verified over `[-126, 127.9)`),
* **99.95 %** of `x` in `[128, 1000)` (step 0.1, all of which should be `+inf`) return NaN or a
  finite wrong value; the only accidentally-correct point is the exact integer `x = 128.0`
  (`f = 0` makes the mantissa zero, so `setexp(1.0, 255)` happens to land on the `+inf` encoding).

## Root cause

`tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_exp2.h`
(and the identical `.../blackhole/...` copy), in `_sfpu_exp2_fp32_accurate_`:

```cpp
r = r * f + sfpi::vConstFloatPrgm1;
y *= std::numeric_limits<float>::infinity();   // <-- correct overflow result y = +inf ...
r = r * f + sfpi::vConstFloatPrgm0;
abs_x = sfpi::abs(x);
r = r * f + 1.0f;                              // r = 2^f

v_if(abs_x >= 0.0f) {
    sfpi::vInt e = sfpi::exexp(r, sfpi::ExponentMode::Biased);
    e += i;
    // e < 255                                 // <-- asserted-but-unenforced invariant
    v_block {
        sfpi::vInt e_lt_255 = __builtin_rvtt_sfpiadd_i(e.get(), -255, sfpi::SFPIADD_MOD1_CC_LT0);
        y = sfpi::setexp(r, e);                // <-- UNCONDITIONAL: clobbers y=+inf when e>=255
        v_if(e_lt_255 < -254) {                // only the underflow (e<1) case is handled
            y = 0.0f;
        }
        v_endif;
    }
    v_endblock;
}
v_endif;
return y;
```

`e_lt_255 = e - 255` is computed but only its `< -254` (underflow) branch is used; the `e >= 255`
(overflow) case falls through to the raw `setexp(r, e)`. Because `setexp` writes only the low 8
bits of the exponent field, `e == 255` with `r`'s nonzero fraction is the canonical NaN pattern,
and `e >= 256` wraps to a small biased exponent (`e & 0xFF`), producing finite garbage.

The default **bf16** path (`_sfpu_exp2_bf16_`) is unaffected — it clamps with
`clamp(xlog2, 0.0f, 255.0f)` and saturates to `+inf` correctly. Only the `is_fp32_dest_acc_en`
path is broken.

## Affected files (both architectures, identical content)

* `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_exp2.h`
* `tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_exp2.h`

## Suggested fix

Gate the `setexp` on the already-computed `e < 255` predicate and let the precomputed
`y = r * inf = +inf` stand for the overflow case (the partial-poly factor is positive on
`f in [-0.5, 0.5]`, so `y` is exactly `+inf`):

```cpp
v_block {
    sfpi::vInt e_lt_255 = __builtin_rvtt_sfpiadd_i(e.get(), -255, sfpi::SFPIADD_MOD1_CC_LT0);
    // Overflow guard (e >= 255): keep the precomputed y = r * inf = +inf.
    v_if(e_lt_255 < 0) {  // e < 255
        y = sfpi::setexp(r, e);
        v_if(e_lt_255 < -254) {   // e < 1: underflow, incl subnormals
            y = 0.0f;
        }
        v_endif;
    }
    v_endif;
}
v_endblock;
```

Full patched file and unified diff attached (`ckernel_sfpu_exp2.patched.h`,
`exp2_overflow_guard.patch`); the change is identical for both the wormhole_b0 and blackhole
copies. No `exp2_init` change is needed.

## Suggested test

Extend the exp2 fp32 accuracy test with the overflow band, comparing against a fp32 golden
(`torch.exp2` / `numpy.exp2`, which saturate to `+inf`):

```python
x = torch.tensor([128.01, 128.25, 129.0, 160.0, 200.0, 256.0, 300.0], dtype=torch.float32)
# fp32 dest-accumulation path
got = ttnn.exp2(x_on_device)               # is_fp32_dest_acc_en=True
assert torch.isinf(got).all()              # currently: NaN / finite garbage
```

The attached `verify_exp2.py` reproduces the whole table and the [128,1000) sweep off-device, and
also validates that the proposed fix restores `+inf` for all `x >= 128` while remaining
bit-identical to the current kernel for every `x < 128`.

## Severity / difficulty

Silent wrong finite results (`exp2(256) = 1.0`) across the entire `x >= 128` half-line on the
fp32-accurate path, with a one-branch fix that reuses a value the kernel already computes.
Comparable in kind to #52037 (logaddexp2 overflow-safe, $1500 medium) and the overflow-
reconstruction genre of #54046 (xielu). Suggested **$1500 (medium)**.

**Comment by 5awtooth** [comment]
I have a complete fix (both Wormhole + Blackhole kernels) plus a regression test and an fp32 numerical proof reproducing the failure and confirming the fix. Requesting assignment so I can open a PR.

**Comment by Luvi-1** [comment]
/attempt #54054

**Plan:** Add an overflow guard in both wormhole_b0 and blackhole `_sfpu_exp2_fp32_accurate_` kernels. The fix wraps `setexp(r, e)` in a `v_if(e_lt_255 < 0)` block so that when the biased exponent >= 255, the already-computed `+inf` (from `y *= inf`) is preserved instead of being overwritten by NaN/wrapped garbage. Adding a regression test covering x in [128, 1000].

**Comment by ncvetkovicTT** [comment]
FYI @jberkowitzTT, let's assess this issue and if valid, assign it. Message to contributors - until you're assigned, don't spam with PRs or comments.

**Comment by jberkowitzTT** [comment]
@5awtooth - please submit any bounty proposals to [bounties@tenstorrent.com](mailto:bounties@tenstorrent.com)

**Comment by jberkowitzTT** [comment]
## 📝 Background

This is a community-suggested bounty that we have reviewed & assigned to @5awtooth 


## 💡 Problem to Solve

>ttnn.exp2: fp32 path returns NaN / finite garbage instead of +inf for x >= 128 (missing overflow guard))


Note: This bounty is beholden to the [Tenstorrent Bounty Program Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html) & local tax requirements may apply

**Comment by jberkowitzTT** [comment]
Removing this from the bounty board
